File: app/src/main/java/tech/salroid/filmy/widget/FilmyWidgetProvider.java
package tech.salroid.filmy.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.widget.RemoteViews;

import com.bumptech.glide.Glide;
import com.bumptech.glide.request.target.AppWidgetTarget;

import tech.salroid.filmy.R;
import tech.salroid.filmy.activities.MovieDetailsActivity;
import tech.salroid.filmy.database.FilmContract;
import tech.salroid.filmy.database.MovieProjection;

/*
 * Filmy Application for Android
 * Copyright (c) 2016 Ramankit Singh (http://github.com/webianks).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class FilmyWidgetProvider extends AppWidgetProvider {

    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {

        final int N = appWidgetIds.length;

        String movie_id=" ";
        String movie_title= " ";
        String movie_poster=" ";

        for (int i=0; i<N; i++) {

            int appWidgetId = appWidgetIds[i];

            Uri moviesForTheUri = FilmContract.MoviesEntry.CONTENT_URI;
            String selection = null;
            String[] selectionArgs = null;
            Cursor cursor = context.getContentResolver().query(moviesForTheUri,MovieProjection.MOVIE_COLUMNS, selection, selectionArgs, null);

            if (cursor != null && cursor.getCount() > 0) {

                cursor.moveToFirst();

                int id_index = cursor.getColumnIndex(FilmContract.MoviesEntry.MOVIE_ID);
                int title_index = cursor.getColumnIndex(FilmContract.MoviesEntry.MOVIE_TITLE);
                int poster_index = cursor.getColumnIndex(FilmContract.MoviesEntry.MOVIE_POSTER_LINK);
                int year_index = cursor.getColumnIndex(FilmContract.MoviesEntry.MOVIE_YEAR);

                movie_id = cursor.getString(id_index);
                movie_title = cursor.getString(title_index);
                movie_poster = cursor.getString(poster_index);
                //String imdb_id = cursor.getString(id_index);
                //int movie_year = cursor.getInt(year_index);



            } else {

            }

            cursor.close();

            final RemoteViews remoteViews = new RemoteViews(context.getPackageName(), R.layout.filmy_appwidget);
            remoteViews.setTextViewText(R.id.widget_movie_name,movie_title);

            AppWidgetTarget appWidgetTarget = new AppWidgetTarget(context, R.id.widget_movie_image, remoteViews, appWidgetIds);

            Glide.with(context)
                    .asBitmap()
                    .load(movie_poster)
                    .into(appWidgetTarget);


            Intent intent = new Intent(context,MovieDetailsActivity.class);
            intent.putExtra("title", movie_title);
            intent.putExtra("activity", true);
            intent.putExtra("type", -1);
            intent.putExtra("database_applicable", false);
            intent.putExtra("network_applicable", true);
            intent.putExtra("id", movie_id);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            PendingIntent pendingIntent = PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);
            remoteViews.setOnClickPendingIntent(R.id.activity_opener, pendingIntent);

            appWidgetManager.updateAppWidget(appWidgetId, remoteViews);

        }
    }



}

File: app/src/main/java/tech/salroid/filmy/ui/widget/FilmyWidgetProvider.kt
package tech.salroid.filmy.ui.widget

import android.app.PendingIntent
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.widget.RemoteViews
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.target.AppWidgetTarget
import kotlinx.coroutines.*
import tech.salroid.filmy.R
import tech.salroid.filmy.ui.details.MovieDetailsActivity
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.DATABASE_APPLICABLE
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.FROM_ACTIVITY
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.MOVIE_ID
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.MOVIE_TITLE
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.MOVIE_TYPE
import tech.salroid.filmy.ui.home.MoviesFragment.Companion.NETWORK_APPLICABLE
import tech.salroid.filmy.ui.home.MoviesRepository

//@AndroidEntryPoint
class FilmyWidgetProvider : AppWidgetProvider() {

    private val job = SupervisorJob()
    private val coroutineScope = CoroutineScope(Dispatchers.IO + job)

    //@Inject
    lateinit var moviesRepository: MoviesRepository

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)

        coroutineScope.launch {
            //val movies = moviesRepository.getTrendingFromLocal()
            val appWidgetManager = AppWidgetManager.getInstance(context)
            val appWidget = AppWidgetManager.getInstance(context)
            val appWidgetIds =
                appWidget.getAppWidgetIds(ComponentName(context, FilmyWidgetProvider::class.java))

           /* if (movies.isNotEmpty() && appWidgetIds.isNotEmpty()) {
                val movie = movies.first()
                updateAppWidget(
                    context,
                    appWidgetManager,
                    appWidgetIds,
                    movie.id,
                    movie.title,
                    movie.backdropPath
                )
            }*/
        }
    }

    override fun onUpdate(
        context: Context?,
        appWidgetManager: AppWidgetManager?,
        appWidgetIds: IntArray?
    ) {
        super.onUpdate(context, appWidgetManager, appWidgetIds)
    }

    private fun updateAppWidget(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray,
        movieId: Int,
        movieTitle: String?,
        moviePoster: String?
    ) {

        //val radius = context.resources.getDimensionPixelSize(R.dimen.filmy8dp)
        val remoteViews = RemoteViews(context.packageName, R.layout.filmy_appwidget).apply {
            setTextViewText(R.id.widget_movie_name, movieTitle)
        }

        val appWidgetTarget = AppWidgetTarget(
            context.applicationContext,
            R.id.widget_movie_image,
            remoteViews,
            *appWidgetIds
        )

        Glide.with(context.applicationContext)
            .asBitmap()
            .diskCacheStrategy(DiskCacheStrategy.NONE)
            //.transform(RoundedCorners(radius))
            .load("https://image.tmdb.org/t/p/w1280$moviePoster")
            .into(appWidgetTarget)

        remoteViews.setOnClickPendingIntent(
            android.R.id.background,
            getPendingIntentActivity(context, movieId, movieTitle)
        )
        appWidgetManager.updateAppWidget(appWidgetIds, remoteViews)
    }

    private fun getPendingIntentActivity(
        context: Context,
        movieId: Int,
        movieTitle: String?
    ): PendingIntent? {
        val intent = Intent(context, MovieDetailsActivity::class.java).apply {
            putExtra(MOVIE_ID, movieId.toString())
            putExtra(MOVIE_TITLE, movieTitle)
            putExtra(FROM_ACTIVITY, true)
            putExtra(MOVIE_TYPE, -1)
            putExtra(DATABASE_APPLICABLE, false)
            putExtra(NETWORK_APPLICABLE, true)
        }
        return PendingIntent.getActivity(
            context,
            0,
            intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
    }

    override fun onDisabled(context: Context?) {
        super.onDisabled(context)
        job.cancel()
    }
}
